src/librustc_mir/hair/mod.rs

```diff
@@ -28,6 +26,8 @@ use self::cx::Cx;
 
 pub mod cx;
 
+pub use rustc_const_eval::pattern::{BindingMode, Pattern, PatternKind, FieldPattern};
```

Comment:
match checking uses a completely different algorithm than match building. HAIR patterns are basically independent of the rest of HAIR.